atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CqlExecutor.java

```diff
@@ -114,19 +116,34 @@ private InetSocketAddress getHostForRow(byte[] row) {
     }
 
     private CqlResult executeQueryOnHost(String query, InetSocketAddress host) {
-        ByteBuffer queryBytes = ByteBuffer.wrap(query.getBytes(StandardCharsets.UTF_8));
-        return executeQueryOnHost(queryBytes, host);
+        return executeCqlFunctionOnHost(getCqlFunction(query), host);
     }
 
-    private CqlResult executeQueryOnHost(ByteBuffer queryBytes, InetSocketAddress host) {
+    private CqlResult executeCqlFunctionOnHost(FunctionCheckedException<Cassandra.Client, CqlResult, TException> fn,
+            InetSocketAddress host) {
         try {
-            return clientPool.runWithRetryOnHost(host, client ->
-                    client.execute_cql3_query(queryBytes, Compression.NONE, consistency));
+            return clientPool.runWithRetryOnHost(host, fn);
         } catch (TException e) {
             throw Throwables.throwUncheckedException(e);
         }
     }
 
+    private FunctionCheckedException<Cassandra.Client, CqlResult, TException> getCqlFunction(String query) {
+        ByteBuffer queryBytes = ByteBuffer.wrap(query.getBytes(StandardCharsets.UTF_8));
```

Comment:
do we want to eagerly capture and retain both the query string & converted bytes & byte buffer or should this be inlined into `apply` below as:

```
            public CqlResult apply(Cassandra.Client client) throws TException {
                ByteBuffer queryBytes = ByteBuffer.wrap(query.getBytes(StandardCharsets.UTF_8));
                return client.execute_cql3_query(queryBytes, Compression.NONE, consistency);
            }
```

Reply:
We need to override `toString` here to make the log message helpful - see `CassandraClientPoolImpl.runWithRetryOnHost`.

Reply:
You can still override toStringas you have it, this just avoids keeping the byte buffer around  after executing the query. If we are going to re-execute the query many times, then your original version of eagerly converting to bytes might be better tradeoff

Reply:
I would expect the query to execute one time in 99.9% of cases.
However, I'm a little confused about the comment about "keeping the byte buffer around" - doesn't it go away when we exit `getCqlFunction`?

Relatedly - I've been biasing towards keeping the stuff inside `clientPool.runWithRetryOnHost` as small as possible, ideally just `client.do_the_thing(the_arguments)`.

That way, it's more transparent which code might throw the exception we're passing, and it generally produces cleaner and more testable code - if we do that for long enough, eventually all of the clientPool-calling logic will live inside a small class which we can just mock for testing (and replace with CQL for production).

Does that practice lead to sub-optimal perf somehow?

Reply:
I don't think this is actually a huge issue as I don't expect the queries to be that large, but my comment was more intended as context around potential issues with capturing references via lambda & anonymous inner classes. By "keeping the byte buffer around", what I mean is that in the current PR, the `ByteBuffer queryBytes` and `String query` references are captured as they are referenced in the `apply` and `toString` methods respectively. This means that both will live beyond the execution of the `getCqlFunction` method and until the `executeCqlFunctionOnHost` exits.

```
    private FunctionCheckedException<Cassandra.Client, CqlResult, TException> getCqlFunction(String query) {
        return new FunctionCheckedException<Cassandra.Client, CqlResult, TException>() {
            @Override
            public CqlResult apply(Cassandra.Client client) throws TException {
                ByteBuffer queryBytes = ByteBuffer.wrap(query.getBytes(StandardCharsets.UTF_8));
                return client.execute_cql3_query(queryBytes, Compression.NONE, consistency);
            }

            @Override
            public String toString() {
                return query;
            }
        };
    }
```